### ISSUE TITLE
Python behind the scenes #6: how Python object system works

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#7434]
Python behind the scenes #6: how Python object system works<br>
<br>
https://ift.tt/3oq13c9<br>
<br>
<br>
<br>
<div>
<div>
<div>
<div>
<p>As we know from the previous parts of this series, the execution of a Python program consists of two major steps:</p>
<ol>
<li>
<a href="https://tenthousandmeters.com/blog/python-behind-the-scenes-2-how-the-cpython-compiler-works/">The CPython compiler</a> translates Python code to bytecode.</li>
<li>
<a href="https://tenthousandmeters.com/blog/python-behind-the-scenes-1-how-the-cpython-vm-works/">The CPython VM</a> executes the bytecode.</li>
</ol>
<p>We've been focusing on the second step for quite a while. In <a href="https://tenthousandmeters.com/blog/python-behind-the-scenes-4-how-python-bytecode-is-executed/">part 4</a> we've looked at the evaluation loop, a place where Python bytecode gets executed. And in <a href="https://tenthousandmeters.com/blog/python-behind-the-scenes-5-how-variables-are-implemented-in-cpython/">part 5</a> we've studied how the VM executes the instructions that are used to implement variables. What we haven't covered yet is how the VM actually computes something. We postponed this question because to answer it, we first need to understand how the most fundamental part of the language works. Today, we'll study the Python object system.</p>
<p><strong>Note</strong>: In this post I'm referring to CPython 3.9. Some implementation details will certainly change as CPython evolves. I'll try to keep track of important changes and add update notes.</p>
<h2>Motivation</h2>
<p>Consider an extremely simple piece of Python code:</p>
<p>To compute the function <code>f</code>, CPython must evaluate the expression <code>x + 7</code>. The question I'd like to ask is: How does CPython do that? Special methods such as <code>__add__()</code> and <code>__radd__()</code> probably come to your mind. When we define these methods on a class, the instances of that class can be added using the <code>+</code> operator. So, you might think that CPython does something like this:</p>
<ol>
<li>It calls <code>x.__add__(7)</code> or <code>type(x).__add__(x, 7)</code>.</li>
<li>If <code>x</code> doesn't have <code>__add__()</code>, or if this method fails, it calls <code>(7).__radd__(x)</code> or <code>int.__radd__(7, x)</code>.</li>
</ol>
<p>The reality, tough, is a bit more complicated. What really happens depends on what <code>x</code> is. For example, if <code>x</code> is an instance of a user-defined class, the algorithm described above resembles the truth. If, however, <code>x</code> is an instance of a built-in type, like <code>int</code> or <code>float</code>, CPython doesn't call any special methods at all.</p>
<p>To learn how some Python code is executed, we can do the following:</p>
<ol>
<li>Disassemble the code into bytecode.</li>
<li>Study how the VM executes the disassembled bytecode instructions.</li>
</ol>
<p>Let's apply this algorithm to the function <code>f</code>. The compiler translates the body of this function to the following bytecode:</p>
<div>
<pre>$ python -m dis f.py
...
  2           0 LOAD_FAST                0 (x)
              2 LOAD_CONST               1 (7)
              4 BINARY_ADD
              6 RETURN_VALUE
</pre>
</div>
<p>And here's what these bytecode instructions do:</p>
<ol>
<li>
<code>LOAD_FAST</code> loads the value of the parameter <code>x</code> onto the stack.</li>
<li>
<code>LOAD_CONST</code> loads the constant <code>7</code> onto the stack.</li>
<li>
<code>BINARY_ADD</code> pops two values from the stack, adds them and pushes the result back onto the stack.</li>
<li>
<code>RETURN_VALUE</code> pops the value from the stack and returns it.</li>
</ol>
<p>How does the VM add two values? To answer this question, we need to understand what these values are. For us, <code>7</code> is an instance of <code>int</code> and <code>x</code> is, well, anything. For the VM, though, everything is a Python object. All values the VM pushes onto the stack and pops from the stack are pointers to <code>PyObject</code> structs (hence the phrase "Everything in Python is an object").</p>
<p>The VM doesn't need to know how to add integers or strings, that is, how to do the arithmetic or concatenate sequences. All it needs to know is that every Python object has a type. A type, in turn, knows everything about its objects. For example, the <code>int</code> type knows how to add integers, and the <code>float</code> type knows how to add floats. So, the VM asks the type to perform the operation.</p>
<p>This simplified explanation captures the essence of the solution, but it also omits a lot of important details. To get a more realistic picture, we need to understand what Python objects and types really are and how they work.</p>
<h2>Python objects and types</h2>
<p>We've discussed Python objects a little in <a href="https://tenthousandmeters.com/blog/python-behind-the-scenes-3-stepping-through-the-cpython-source-code/">part 3</a>. This discussion is worth repeating here.</p>
<p>We begin with the definition of the <code>PyObject</code> struct:</p>
<div>
<pre><span>typedef</span> <span>struct</span> <span>_object</span> <span>{</span>
    <span>_PyObject_HEAD_EXTRA</span> <span>// macro, for debugging purposes only</span>
    <span>Py_ssize_t</span> <span>ob_refcnt</span><span>;</span>
    <span>PyTypeObject</span> <span>*</span><span>ob_type</span><span>;</span>
<span>}</span> <span>PyObject</span><span>;</span>
</pre>
</div>
<p>It has two members:</p>
<ul>
<li>a reference count <code>ob_refcnt</code> that CPython uses for garbage collection; and</li>
<li>a pointer to the object's type <code>ob_type</code>.</li>
</ul>
<p>We said that the VM treats any Python object as <code>PyObject</code>. How is that possible? The C programming language has no notion of classes and inheritance. Nevertheless, it's possible to implement in C something that can be called a single inheritance. The C standard states that a pointer to any struct can be converted to a pointer to its first member and vice versa. So, we can "extend" <code>PyObject</code> by defining a new struct whose first member is <code>PyObject</code>.</p>
<p>Here's, for example, how the <code>float</code> object is defined:</p>
<div>
<pre><span>typedef</span> <span>struct</span> <span>{</span>
    <span>PyObject</span> <span>ob_base</span><span>;</span> <span>// expansion of PyObject_HEAD macro</span>
    <span>double</span> <span>ob_fval</span><span>;</span>
<span>}</span> <span>PyFloatObject</span><span>;</span>
</pre>
</div>
<p>A <code>float</code> object stores everything <code>PyObject</code> stores plus a floating-point value <code>ob_fval</code>. The C standard simply states that we can convert a pointer to <code>PyFloatObject</code> to a pointer to <code>PyObject</code> and vice versa:</p>
<div>
<pre><span>PyFloatObject</span> <span>float_object</span><span>;</span>
<span>// ...</span>
<span>PyObject</span> <span>*</span><span>obj_ptr</span> <span>=</span> <span>(</span><span>PyObject</span> <span>*</span><span>)</span><span>&amp;</span><span>float_object</span><span>;</span>
<span>PyFloatObject</span> <span>*</span><span>float_obj_ptr</span> <span>=</span> <span>(</span><span>PyFloatObject</span> <span>*</span><span>)</span><span>obj_ptr</span><span>;</span>
</pre>
</div>
<p>The reason why the VM treats every Python object as <code>PyObject</code> is because all it needs to access is the object's type. A type is also a Python object, an instance of the <code>PyTypeObject</code> struct:</p>
<div>
<pre><span>// PyTypeObject is a typedef for "struct _typeobject"</span>

<span>struct</span> <span>_typeobject</span> <span>{</span>
    <span>PyVarObject</span> <span>ob_base</span><span>;</span> <span>// expansion of PyObject_VAR_HEAD macro</span>
    <span>const</span> <span>char</span> <span>*</span><span>tp_name</span><span>;</span> <span>/* For printing, in format "&lt;module&gt;.&lt;name&gt;" */</span>
    <span>Py_ssize_t</span> <span>tp_basicsize</span><span>,</span> <span>tp_itemsize</span><span>;</span> <span>/* For allocation */</span>

    <span>/* Methods to implement standard operations */</span>

    <span>destructor</span> <span>tp_dealloc</span><span>;</span>
    <span>Py_ssize_t</span> <span>tp_vectorcall_offset</span><span>;</span>
    <span>getattrfunc</span> <span>tp_getattr</span><span>;</span>
    <span>setattrfunc</span> <span>tp_setattr</span><span>;</span>
    <span>PyAsyncMethods</span> <span>*</span><span>tp_as_async</span><span>;</span> <span>/* formerly known as tp_compare (Python 2)</span>
<span> or tp_reserved (Python 3) */</span>
    <span>reprfunc</span> <span>tp_repr</span><span>;</span>

    <span>/* Method suites for standard classes */</span>

    <span>PyNumberMethods</span> <span>*</span><span>tp_as_number</span><span>;</span>
    <span>PySequenceMethods</span> <span>*</span><span>tp_as_sequence</span><span>;</span>
    <span>PyMappingMethods</span> <span>*</span><span>tp_as_mapping</span><span>;</span>

    <span>/* More standard operations (here for binary compatibility) */</span>

    <span>hashfunc</span> <span>tp_hash</span><span>;</span>
    <span>ternaryfunc</span> <span>tp_call</span><span>;</span>
    <span>reprfunc</span> <span>tp_str</span><span>;</span>
    <span>getattrofunc</span> <span>tp_getattro</span><span>;</span>
    <span>setattrofunc</span> <span>tp_setattro</span><span>;</span>

    <span>/* Functions to access object as input/output buffer */</span>
    <span>PyBufferProcs</span> <span>*</span><span>tp_as_buffer</span><span>;</span>

    <span>/* Flags to define presence of optional/expanded features */</span>
    <span>unsigned</span> <span>long</span> <span>tp_flags</span><span>;</span>

    <span>const</span> <span>char</span> <span>*</span><span>tp_doc</span><span>;</span> <span>/* Documentation string */</span>

    <span>/* Assigned meaning in release 2.0 */</span>
    <span>/* call function for all accessible objects */</span>
    <span>traverseproc</span> <span>tp_traverse</span><span>;</span>

    <span>/* delete references to contained objects */</span>
    <span>inquiry</span> <span>tp_clear</span><span>;</span>

    <span>/* Assigned meaning in release 2.1 */</span>
    <span>/* rich comparisons */</span>
    <span>richcmpfunc</span> <span>tp_richcompare</span><span>;</span>

    <span>/* weak reference enabler */</span>
    <span>Py_ssize_t</span> <span>tp_weaklistoffset</span><span>;</span>

    <span>/* Iterators */</span>
    <span>getiterfunc</span> <span>tp_iter</span><span>;</span>
    <span>iternextfunc</span> <span>tp_iternext</span><span>;</span>

    <span>/* Attribute descriptor and subclassing stuff */</span>
    <span>struct</span> <span>PyMethodDef</span> <span>*</span><span>tp_methods</span><span>;</span>
    <span>struct</span> <span>PyMemberDef</span> <span>*</span><span>tp_members</span><span>;</span>
    <span>struct</span> <span>PyGetSetDef</span> <span>*</span><span>tp_getset</span><span>;</span>
    <span>struct</span> <span>_typeobject</span> <span>*</span><span>tp_base</span><span>;</span>
    <span>PyObject</span> <span>*</span><span>tp_dict</span><span>;</span>
    <span>descrgetfunc</span> <span>tp_descr_get</span><span>;</span>
    <span>descrsetfunc</span> <span>tp_descr_set</span><span>;</span>
    <span>Py_ssize_t</span> <span>tp_dictoffset</span><span>;</span>
    <span>initproc</span> <span>tp_init</span><span>;</span>
    <span>allocfunc</span> <span>tp_alloc</span><span>;</span>
    <span>newfunc</span> <span>tp_new</span><span>;</span>
    <span>freefunc</span> <span>tp_free</span><span>;</span> <span>/* Low-level free-memory routine */</span>
    <span>inquiry</span> <span>tp_is_gc</span><span>;</span> <span>/* For PyObject_IS_GC */</span>
    <span>PyObject</span> <span>*</span><span>tp_bases</span><span>;</span>
    <span>PyObject</span> <span>*</span><span>tp_mro</span><span>;</span> <span>/* method resolution order */</span>
    <span>PyObject</span> <span>*</span><span>tp_cache</span><span>;</span>
    <span>PyObject</span> <span>*</span><span>tp_subclasses</span><span>;</span>
    <span>PyObject</span> <span>*</span><span>tp_weaklist</span><span>;</span>
    <span>destructor</span> <span>tp_del</span><span>;</span>

    <span>/* Type attribute cache version tag. Added in version 2.6 */</span>
    <span>unsigned</span> <span>int</span> <span>tp_version_tag</span><span>;</span>

    <span>destructor</span> <span>tp_finalize</span><span>;</span>
    <span>vectorcallfunc</span> <span>tp_vectorcall</span><span>;</span>
<span>};</span>
</pre>
</div>
<p>By the way, note that the first member of a type is not <code>PyObject</code> but <code>PyVarObject</code>, which is defined as follows:</p>
<div>
<pre><span>typedef</span> <span>struct</span> <span>{</span>
    <span>PyObject</span> <span>ob_base</span><span>;</span>
    <span>Py_ssize_t</span> <span>ob_size</span><span>;</span> <span>/* Number of items in variable part */</span>
<span>}</span> <span>PyVarObject</span><span>;</span>
</pre>
</div>
<p>Nevertheless, since the first member of <code>PyVarObject</code> is <code>PyObject</code>, a pointer to a type can still be converted to a pointer to <code>PyObject</code>.</p>
<p>So, what is a type and why does it have so many members? A type determines how the objects of that type behave. Each member of a type, called slot, is responsible for a particular aspect of the object's behavior. For example:</p>
<ul>
<li>
<code>tp_new</code> is a pointer to a function that creates new objects of the type.</li>
<li>
<code>tp_str</code> is a pointer to a function that implements <code>str()</code> for objects of the type.</li>
<li>
<code>tp_hash</code> is a pointer to a function that implements <code>hash()</code> for objects of the type.</li>
</ul>
<p>Some slots, called sub-slots, are grouped together in suites. A suite is just a struct that contains related slots. For example, the <code>PySequenceMethods</code> struct is a suite of sub-slots that implement the sequence protocol:</p>
<div>
<pre><span>typedef</span> <span>struct</span> <span>{</span>
    <span>lenfunc</span> <span>sq_length</span><span>;</span>
    <span>binaryfunc</span> <span>sq_concat</span><span>;</span>
    <span>ssizeargfunc</span> <span>sq_repeat</span><span>;</span>
    <span>ssizeargfunc</span> <span>sq_item</span><span>;</span>
    <span>void</span> <span>*</span><span>was_sq_slice</span><span>;</span>
    <span>ssizeobjargproc</span> <span>sq_ass_item</span><span>;</span>
    <span>void</span> <span>*</span><span>was_sq_ass_slice</span><span>;</span>
    <span>objobjproc</span> <span>sq_contains</span><span>;</span>

    <span>binaryfunc</span> <span>sq_inplace_concat</span><span>;</span>
    <span>ssizeargfunc</span> <span>sq_inplace_repeat</span><span>;</span>
<span>}</span> <span>PySequenceMethods</span><span>;</span>
</pre>
</div>
<p>If you count all the slots and sub-slots, you'll get a scary number. Fortunately, each slot is very well <a href="https://docs.python.org/3/c-api/typeobj.html">documented</a> in the Python/C API Reference Manual (I strongly recommend you to bookmark this link). Today we'll cover only a few slots. Nevertheless, it shall give us a general idea of how slots are used.</p>
<p>Since we're interested in how CPython adds objects, let's find the slots responsible for addition. There must be at least one such slot. After careful inspection of the <code>PyTypeObject</code> struct, we find that it has the "number" suite <code>PyNumberMethods</code>, and the first slot of this suite is a binary function called <code>nd_add</code>:</p>
<div>
<pre><span>typedef</span> <span>struct</span> <span>{</span>
    <span>binaryfunc</span> <span>nb_add</span><span>;</span> <span>// typedef PyObject * (*binaryfunc)(PyObject *, PyObject *)</span>
    <span>binaryfunc</span> <span>nb_subtract</span><span>;</span>
    <span>binaryfunc</span> <span>nb_multiply</span><span>;</span>
    <span>binaryfunc</span> <span>nb_remainder</span><span>;</span>
    <span>binaryfunc</span> <span>nb_divmod</span><span>;</span>
    <span>// ... more sub-slots</span>
<span>}</span> <span>PyNumberMethods</span><span>;</span>
</pre>
</div>
<p>It seems that the <code>nb_add</code> slot is what we're looking for. Two questions naturally arise regarding this slot:</p>
<ul>
<li>
<p>What is it set to?</p>
</li>
<li>
<p>How is it used?</p>
</li>
</ul>
<p>I think it's better to start with the second. We should expect that the VM calls <code>nb_add</code> to execute the <code>BINARY_ADD</code> opcode. So, let's, for a moment, suspend our discussion about types and take a look at how the <code>BINARY_ADD</code> opcode is implemented.</p>
<h2>BINARY_ADD</h2>
<p>Like any other opcode, <code>BINARY_ADD</code> is implemented in the evaluation loop in <a href="https://github.com/python/cpython/blob/3.9/Python/ceval.c#L1684"><code>Python/ceval.c</code></a>:</p>
<div>
<pre><span>case</span> <span>TARGET</span><span>(</span><span>BINARY_ADD</span><span>)</span><span>:</span> <span>{</span>
    <span>PyObject</span> <span>*</span><span>right</span> <span>=</span> <span>POP</span><span>();</span>
    <span>PyObject</span> <span>*</span><span>left</span> <span>=</span> <span>TOP</span><span>();</span>
    <span>PyObject</span> <span>*</span><span>sum</span><span>;</span>
    <span>/* NOTE(haypo): Please don't try to micro-optimize int+int on</span>
<span> CPython using bytecode, it is simply worthless.</span>
<span> See http://bugs.python.org/issue21955 and</span>
<span> http://bugs.python.org/issue10044 for the discussion. In short,</span>
<span> no patch shown any impact on a realistic benchmark, only a minor</span>
<span> speedup on microbenchmarks. */</span>
    <span>if</span> <span>(</span><span>PyUnicode_CheckExact</span><span>(</span><span>left</span><span>)</span> <span>&amp;&amp;</span>
                <span>PyUnicode_CheckExact</span><span>(</span><span>right</span><span>))</span> <span>{</span>
        <span>sum</span> <span>=</span> <span>unicode_concatenate</span><span>(</span><span>tstate</span><span>,</span> <span>left</span><span>,</span> <span>right</span><span>,</span> <span>f</span><span>,</span> <span>next_instr</span><span>);</span>
        <span>/* unicode_concatenate consumed the ref to left */</span>
    <span>}</span>
    <span>else</span> <span>{</span>
        <span>sum</span> <span>=</span> <span>PyNumber_Add</span><span>(</span><span>left</span><span>,</span> <span>right</span><span>);</span>
        <span>Py_DECREF</span><span>(</span><span>left</span><span>);</span>
    <span>}</span>
    <span>Py_DECREF</span><span>(</span><span>right</span><span>);</span>
    <span>SET_TOP</span><span>(</span><span>sum</span><span>);</span>
    <span>if</span> <span>(</span><span>sum</span> <span>==</span> <span>NULL</span><span>)</span>
        <span>goto</span> <span>error</span><span>;</span>
    <span>DISPATCH</span><span>();</span>
<span>}</span>
</pre>
</div>
<p>This code requires some comments. We can see that it calls <code>PyNumber_Add()</code> to add two objects, but if the objects are strings, it calls <code>unicode_concatenate()</code> instead. Why so? This is an optimization. Python strings seem immutable, but sometimes CPython mutates a string and thus avoids creating a new string. Consider appending one string to another:</p>
<p>If the <code>output</code> variable points to a string that has no other references, it's safe to mutate that string. This is exactly the logic that <code>unicode_concatenate()</code> implements.</p>
<p>It might be tempting to handle other special cases in the evaluation loop as well and optimize, for example, integers and floats. The comment explicitly warns against it. The problem is that a new special case comes with an additional check, and this check is only useful when it succeeds. Otherwise, it may have a negative effect on performance.</p>
<p>After this little digression, let's look at <code>PyNumber_Add()</code>:</p>
<div>
<pre><span>PyObject</span> <span>*</span>
<span>PyNumber_Add</span><span>(</span><span>PyObject</span> <span>*</span><span>v</span><span>,</span> <span>PyObject</span> <span>*</span><span>w</span><span>)</span>
<span>{</span>
    <span>// NB_SLOT(nb_add) expands to "offsetof(PyNumberMethods, nb_add)"</span>
    <span>PyObject</span> <span>*</span><span>result</span> <span>=</span> <span>binary_op1</span><span>(</span><span>v</span><span>,</span> <span>w</span><span>,</span> <span>NB_SLOT</span><span>(</span><span>nb_add</span><span>));</span>
    <span>if</span> <span>(</span><span>result</span> <span>==</span> <span>Py_NotImplemented</span><span>)</span> <span>{</span>
        <span>PySequenceMethods</span> <span>*</span><span>m</span> <span>=</span> <span>Py_TYPE</span><span>(</span><span>v</span><span>)</span><span>-&gt;</span><span>tp_as_sequence</span><span>;</span>
        <span>Py_DECREF</span><span>(</span><span>result</span><span>);</span>
        <span>if</span> <span>(</span><span>m</span> <span>&amp;&amp;</span> <span>m</span><span>-&gt;</span><span>sq_concat</span><span>)</span> <span>{</span>
            <span>return</span> <span>(</span><span>*</span><span>m</span><span>-&gt;</span><span>sq_concat</span><span>)(</span><span>v</span><span>,</span> <span>w</span><span>);</span>
        <span>}</span>
        <span>result</span> <span>=</span> <span>binop_type_error</span><span>(</span><span>v</span><span>,</span> <span>w</span><span>,</span> <span>"+"</span><span>);</span>
    <span>}</span>
    <span>return</span> <span>result</span><span>;</span>
<span>}</span>
</pre>
</div>
<p>I suggest to step into <code>binary_op1()</code> straight away and figure out what the rest of <code>PyNumber_Add()</code> does later:</p>
<div>
<pre><span>static</span> <span>PyObject</span> <span>*</span>
<span>binary_op1</span><span>(</span><span>PyObject</span> <span>*</span><span>v</span><span>,</span> <span>PyObject</span> <span>*</span><span>w</span><span>,</span> <span>const</span> <span>int</span> <span>op_slot</span><span>)</span>
<span>{</span>
    <span>PyObject</span> <span>*</span><span>x</span><span>;</span>
    <span>binaryfunc</span> <span>slotv</span> <span>=</span> <span>NULL</span><span>;</span>
    <span>binaryfunc</span> <span>slotw</span> <span>=</span> <span>NULL</span><span>;</span>

    <span>if</span> <span>(</span><span>Py_TYPE</span><span>(</span><span>v</span><span>)</span><span>-&gt;</span><span>tp_as_number</span> <span>!=</span> <span>NULL</span><span>)</span>
        <span>slotv</span> <span>=</span> <span>NB_BINOP</span><span>(</span><span>Py_TYPE</span><span>(</span><span>v</span><span>)</span><span>-&gt;</span><span>tp_as_number</span><span>,</span> <span>op_slot</span><span>);</span>
    <span>if</span> <span>(</span><span>!</span><span>Py_IS_TYPE</span><span>(</span><span>w</span><span>,</span> <span>Py_TYPE</span><span>(</span><span>v</span><span>))</span> <span>&amp;&amp;</span>
        <span>Py_TYPE</span><span>(</span><span>w</span><span>)</span><span>-&gt;</span><span>tp_as_number</span> <span>!=</span> <span>NULL</span><span>)</span> <span>{</span>
        <span>slotw</span> <span>=</span> <span>NB_BINOP</span><span>(</span><span>Py_TYPE</span><span>(</span><span>w</span><span>)</span><span>-&gt;</span><span>tp_as_number</span><span>,</span> <span>op_slot</span><span>);</span>
        <span>if</span> <span>(</span><span>slotw</span> <span>==</span> <span>slotv</span><span>)</span>
            <span>slotw</span> <span>=</span> <span>NULL</span><span>;</span>
    <span>}</span>
    <span>if</span> <span>(</span><span>slotv</span><span>)</span> <span>{</span>
        <span>if</span> <span>(</span><span>slotw</span> <span>&amp;&amp;</span> <span>PyType_IsSubtype</span><span>(</span><span>Py_TYPE</span><span>(</span><span>w</span><span>),</span> <span>Py_TYPE</span><span>(</span><span>v</span><span>)))</span> <span>{</span>
            <span>x</span> <span>=</span> <span>slotw</span><span>(</span><span>v</span><span>,</span> <span>w</span><span>);</span>
            <span>if</span> <span>(</span><span>x</span> <span>!=</span> <span>Py_NotImplemented</span><span>)</span>
                <span>return</span> <span>x</span><span>;</span>
            <span>Py_DECREF</span><span>(</span><span>x</span><span>);</span> <span>/* can't do it */</span>
            <span>slotw</span> <span>=</span> <span>NULL</span><span>;</span>
        <span>}</span>
        <span>x</span> <span>=</span> <span>slotv</span><span>(</span><span>v</span><span>,</span> <span>w</span><span>);</span>
        <span>if</span> <span>(</span><span>x</span> <span>!=</span> <span>Py_NotImplemented</span><span>)</span>
            <span>return</span> <span>x</span><span>;</span>
        <span>Py_DECREF</span><span>(</span><span>x</span><span>);</span> <span>/* can't do it */</span>
    <span>}</span>
    <span>if</span> <span>(</span><span>slotw</span><span>)</span> <span>{</span>
        <span>x</span> <span>=</span> <span>slotw</span><span>(</span><span>v</span><span>,</span> <span>w</span><span>);</span>
        <span>if</span> <span>(</span><span>x</span> <span>!=</span> <span>Py_NotImplemented</span><span>)</span>
            <span>return</span> <span>x</span><span>;</span>
        <span>Py_DECREF</span><span>(</span><span>x</span><span>);</span> <span>/* can't do it */</span>
    <span>}</span>
    <span>Py_RETURN_NOTIMPLEMENTED</span><span>;</span>
<span>}</span>
</pre>
</div>
<p>The <code>binary_op1()</code> function takes three parameters: the left operand, the right operand and an offset that identifies the slot. Types of both operands can implement the slot. Therefore, <code>binary_op1()</code> looks up both implementations. To calculate the result, it calls one implementation or another relying on the following logic:</p>
<ol>
<li>
<p>If the type of one operand is a subtype of another, call the slot of the subtype.</p>
</li>
<li>
<p>If the left operand doesn't have the slot, call the slot of the right operand.</p>
</li>
<li>
<p>Otherwise, call the slot of the left operand.</p>
</li>
</ol>
<p>The reason to prioritize the slot of a subtype is to allow the subtypes to override the behavior of their ancestors:</p>
<div>
<pre><span>$ python -q</span>
<span>&gt;&gt;&gt; </span><span>class</span> <span>HungryInt</span><span>(</span><span>int</span><span>):</span>
<span>... </span>    <span>def</span> <span>__add__</span><span>(</span><span>self</span><span>,</span> <span>o</span><span>):</span>
<span>... </span>            <span>return</span> <span>self</span>
<span>...</span>
<span>&gt;&gt;&gt; </span><span>x</span> <span>=</span> <span>HungryInt</span><span>(</span><span>5</span><span>)</span>
<span>&gt;&gt;&gt; </span><span>x</span> <span>+</span> <span>2</span>
<span>5</span>
<span>&gt;&gt;&gt; </span><span>2</span> <span>+</span> <span>x</span>
<span>7</span>
<span>&gt;&gt;&gt; </span><span>HungryInt</span><span>.</span><span>__radd__</span> <span>=</span> <span>lambda</span> <span>self</span><span>,</span> <span>o</span><span>:</span> <span>self</span>
<span>&gt;&gt;&gt; </span><span>2</span> <span>+</span> <span>x</span>
<span>5</span>
</pre>
</div>
<p>Let's turn back to <code>PyNumber_Add()</code>. If <code>binary_op1()</code> succeeds, <code>PyNumber_Add()</code> simply returns the result of <code>binary_op1()</code>. If, however, <code>binary_op1()</code> returns the <a href="https://docs.python.org/3/library/constants.html#NotImplemented"><code>NotImplemented</code></a> constant, which means that the operation cannot be performed for a given combination of types, <code>PyNumber_Add()</code> calls the <code>sq_concat</code> "sequence" slot of the first operand and returns the result of this call:</p>
<div>
<pre><span>PySequenceMethods</span> <span>*</span><span>m</span> <span>=</span> <span>Py_TYPE</span><span>(</span><span>v</span><span>)</span><span>-&gt;</span><span>tp_as_sequence</span><span>;</span>
<span>if</span> <span>(</span><span>m</span> <span>&amp;&amp;</span> <span>m</span><span>-&gt;</span><span>sq_concat</span><span>)</span> <span>{</span>
    <span>return</span> <span>(</span><span>*</span><span>m</span><span>-&gt;</span><span>sq_concat</span><span>)(</span><span>v</span><span>,</span> <span>w</span><span>);</span>
<span>}</span>
</pre>
</div>
<p>A type can support the <code>+</code> operator either by implementing <code>nb_add</code> or <code>sq_concat</code>. These slots have different meanings:</p>
<ul>
<li>
<code>nb_add</code> means algebraic addition with properties like <code>a + b = b + a</code>.</li>
<li>
<code>sq_concat</code> means the concatenation of sequences.</li>
</ul>
<p>Built-in types such as <code>int</code> and <code>float</code> implement <code>nb_add</code>, and built-in types such as <code>str</code> and <code>list</code> implement <code>sq_concat</code>. Technically, there's no much difference. The main reason to choose one slot over another is to indicate the appropriate meaning. In fact, the <code>sq_concat</code> slot is so unnecessary that it's set to <code>NULL</code> for all user-defined types (i.e. classes).</p>
<p>We saw how the <code>nb_add</code> slot is used: it's called by the <code>binary_op1()</code> function. The next step is to see what it's set to.</p>
<h2>What nb_add can be</h2>
<p>Since addition is a different operation for different types, the <code>nb_add</code> slot of a type must be one of two things:</p>
<ul>
<li>it's either a type-specific function that adds object of that type; or</li>
<li>it's a type-agnostic function that calls some type-specific functions, such as type's <code>__add__()</code> special method.</li>
</ul>
<p>It's indeed one of these two, and which one depends on the type. For example, built-in types such as <code>int</code> and <code>float</code> have their own implementations of <code>nb_add</code>. In contrast, all classes share the same implementation. Fundamentally, built-in types and classes are the same thing &ndash;&nbsp;instances of <code>PyTypeObject</code>. The important difference between them is how they are created. This difference effects the way the slots are set, so we should discuss it.</p>
<h2>Ways to create a type</h2>
<p>There are two ways to create a type object:</p>
<ul>
<li>by statically defining it; or</li>
<li>by dynamically allocating it.</li>
</ul>
<h3>Statically defined types</h3>
<p>An example of a statically defined type is any built-in type. Here's, for instance, how CPython defines the <code>float</code> type:</p>
<div>
<pre><span>PyTypeObject</span> <span>PyFloat_Type</span> <span>=</span> <span>{</span>
    <span>PyVarObject_HEAD_INIT</span><span>(</span><span>&amp;</span><span>PyType_Type</span><span>,</span> <span>0</span><span>)</span>
    <span>"float"</span><span>,</span>
    <span>sizeof</span><span>(</span><span>PyFloatObject</span><span>),</span>
    <span>0</span><span>,</span>
    <span>(</span><span>destructor</span><span>)</span><span>float_dealloc</span><span>,</span>                  <span>/* tp_dealloc */</span>
    <span>0</span><span>,</span>                                          <span>/* tp_vectorcall_offset */</span>
    <span>0</span><span>,</span>                                          <span>/* tp_getattr */</span>
    <span>0</span><span>,</span>                                          <span>/* tp_setattr */</span>
    <span>0</span><span>,</span>                                          <span>/* tp_as_async */</span>
    <span>(</span><span>reprfunc</span><span>)</span><span>float_repr</span><span>,</span>                       <span>/* tp_repr */</span>
    <span>&amp;</span><span>float_as_number</span><span>,</span>                           <span>/* tp_as_number */</span>
    <span>0</span><span>,</span>                                          <span>/* tp_as_sequence */</span>
    <span>0</span><span>,</span>                                          <span>/* tp_as_mapping */</span>
    <span>(</span><span>hashfunc</span><span>)</span><span>float_hash</span><span>,</span>                       <span>/* tp_hash */</span>
    <span>0</span><span>,</span>                                          <span>/* tp_call */</span>
    <span>0</span><span>,</span>                                          <span>/* tp_str */</span>
    <span>PyObject_GenericGetAttr</span><span>,</span>                    <span>/* tp_getattro */</span>
    <span>0</span><span>,</span>                                          <span>/* tp_setattro */</span>
    <span>0</span><span>,</span>                                          <span>/* tp_as_buffer */</span>
    <span>Py_TPFLAGS_DEFAULT</span> <span>|</span> <span>Py_TPFLAGS_BASETYPE</span><span>,</span>   <span>/* tp_flags */</span>
    <span>float_new__doc__</span><span>,</span>                           <span>/* tp_doc */</span>
    <span>0</span><span>,</span>                                          <span>/* tp_traverse */</span>
    <span>0</span><span>,</span>                                          <span>/* tp_clear */</span>
    <span>float_richcompare</span><span>,</span>                          <span>/* tp_richcompare */</span>
    <span>0</span><span>,</span>                                          <span>/* tp_weaklistoffset */</span>
    <span>0</span><span>,</span>                                          <span>/* tp_iter */</span>
    <span>0</span><span>,</span>                                          <span>/* tp_iternext */</span>
    <span>float_methods</span><span>,</span>                              <span>/* tp_methods */</span>
    <span>0</span><span>,</span>                                          <span>/* tp_members */</span>
    <span>float_getset</span><span>,</span>                               <span>/* tp_getset */</span>
    <span>0</span><span>,</span>                                          <span>/* tp_base */</span>
    <span>0</span><span>,</span>                                          <span>/* tp_dict */</span>
    <span>0</span><span>,</span>                                          <span>/* tp_descr_get */</span>
    <span>0</span><span>,</span>                                          <span>/* tp_descr_set */</span>
    <span>0</span><span>,</span>                                          <span>/* tp_dictoffset */</span>
    <span>0</span><span>,</span>                                          <span>/* tp_init */</span>
    <span>0</span><span>,</span>                                          <span>/* tp_alloc */</span>
    <span>float_new</span><span>,</span>                                  <span>/* tp_new */</span>
<span>};</span>
</pre>
</div>
<p>The slots of a statically defined type are specified explicitly. We can easily see how the <code>float</code> type implements <code>nb_add</code> by looking at the "number" suite:</p>
<div>
<pre><span>static</span> <span>PyNumberMethods</span> <span>float_as_number</span> <span>=</span> <span>{</span>
    <span>float_add</span><span>,</span>          <span>/* nb_add */</span>
    <span>float_sub</span><span>,</span>          <span>/* nb_subtract */</span>
    <span>float_mul</span><span>,</span>          <span>/* nb_multiply */</span>
    <span>// ... more number slots</span>
<span>};</span>
</pre>
</div>
<p>where we find the <code>float_add()</code> function, a straightforward implementation of <code>nb_add</code>:</p>
<div>
<pre><span>static</span> <span>PyObject</span> <span>*</span>
<span>float_add</span><span>(</span><span>PyObject</span> <span>*</span><span>v</span><span>,</span> <span>PyObject</span> <span>*</span><span>w</span><span>)</span>
<span>{</span>
    <span>double</span> <span>a</span><span>,</span><span>b</span><span>;</span>
    <span>CONVERT_TO_DOUBLE</span><span>(</span><span>v</span><span>,</span> <span>a</span><span>);</span>
    <span>CONVERT_TO_DOUBLE</span><span>(</span><span>w</span><span>,</span> <span>b</span><span>);</span>
    <span>a</span> <span>=</span> <span>a</span> <span>+</span> <span>b</span><span>;</span>
    <span>return</span> <span>PyFloat_FromDouble</span><span>(</span><span>a</span><span>);</span>
<span>}</span>
</pre>
</div>
<p>The floating-point arithmetic is not that important for our discussion. This example demonstrates how to specify the behavior of a statically defined type. It turned out to be quite easy: just write the implementation of slots and point each slot to the corresponding implementation.</p>
<p>If you want to learn how to statically define your own types, check out <a href="https://docs.python.org/3/extending/newtypes_tutorial.html">Python's tutorial for C/C++ programmers</a>.</p>
<h3>Dynamically allocated types</h3>
<p>Dynamically allocated types are the types we define using the <code>class</code> statement. As we've already said, they are instances of <code>PyTypeObject</code>, just like statically defined types. Traditionally, we call them classes but we might call them user-defined types as well.</p>
<p>From the programmer's perspective, it's easier to define a class in Python than a type in C. This is because CPython does a lot of things behind the scenes when it creates a class. Let's see what's involved in this process.</p>
<p>If we wouldn't know where to start, we could apply the familiar method:</p>
<p>&nbsp;1. Define a simple class</p>
<p>&nbsp;2. Run the disassembler:</p>
<div>
<pre>$ python -m dis class_A.py
</pre>
</div>
<p>&nbsp;3. Study how the VM executes the produced bytecode instructions.</p>
<p>Feel free to do that if you find the time, or read <a href="https://eli.thegreenplace.net/2012/06/15/under-the-hood-of-python-class-definitions">the article on classes</a> by Eli Bendersky. We'll take a shortcut.</p>
<p>An object is created by a call to a type, e.g. <code>list()</code> or <code>MyClass()</code>. A class is created by a call to a metatype. A metatype is just a type whose instances are types. Python has one built-in metatype called <code>PyType_Type</code>, which is known to us simply as <code>type</code>. Here's how it's defined:</p>
<div>
<pre><span>PyTypeObject</span> <span>PyType_Type</span> <span>=</span> <span>{</span>
    <span>PyVarObject_HEAD_INIT</span><span>(</span><span>&amp;</span><span>PyType_Type</span><span>,</span> <span>0</span><span>)</span>
    <span>"type"</span><span>,</span>                                     <span>/* tp_name */</span>
    <span>sizeof</span><span>(</span><span>PyHeapTypeObject</span><span>),</span>                   <span>/* tp_basicsize */</span>
    <span>sizeof</span><span>(</span><span>PyMemberDef</span><span>),</span>                        <span>/* tp_itemsize */</span>
    <span>(</span><span>destructor</span><span>)</span><span>type_dealloc</span><span>,</span>                   <span>/* tp_dealloc */</span>
    <span>offsetof</span><span>(</span><span>PyTypeObject</span><span>,</span> <span>tp_vectorcall</span><span>),</span>      <span>/* tp_vectorcall_offset */</span>
    <span>0</span><span>,</span>                                          <span>/* tp_getattr */</span>
    <span>0</span><span>,</span>                                          <span>/* tp_setattr */</span>
    <span>0</span><span>,</span>                                          <span>/* tp_as_async */</span>
    <span>(</span><span>reprfunc</span><span>)</span><span>type_repr</span><span>,</span>                        <span>/* tp_repr */</span>
    <span>0</span><span>,</span>                                          <span>/* tp_as_number */</span>
    <span>0</span><span>,</span>                                          <span>/* tp_as_sequence */</span>
    <span>0</span><span>,</span>                                          <span>/* tp_as_mapping */</span>
    <span>0</span><span>,</span>                                          <span>/* tp_hash */</span>
    <span>(</span><span>ternaryfunc</span><span>)</span><span>type_call</span><span>,</span>                     <span>/* tp_call */</span>
    <span>0</span><span>,</span>                                          <span>/* tp_str */</span>
    <span>(</span><span>getattrofunc</span><span>)</span><span>type_getattro</span><span>,</span>                <span>/* tp_getattro */</span>
    <span>(</span><span>setattrofunc</span><span>)</span><span>type_setattro</span><span>,</span>                <span>/* tp_setattro */</span>
    <span>0</span><span>,</span>                                          <span>/* tp_as_buffer */</span>
    <span>Py_TPFLAGS_DEFAULT</span> <span>|</span> <span>Py_TPFLAGS_HAVE_GC</span> <span>|</span>
    <span>Py_TPFLAGS_BASETYPE</span> <span>|</span> <span>Py_TPFLAGS_TYPE_SUBCLASS</span> <span>|</span>
    <span>Py_TPFLAGS_HAVE_VECTORCALL</span><span>,</span>                 <span>/* tp_flags */</span>
    <span>type_doc</span><span>,</span>                                   <span>/* tp_doc */</span>
    <span>(</span><span>traverseproc</span><span>)</span><span>type_traverse</span><span>,</span>                <span>/* tp_traverse */</span>
    <span>(</span><span>inquiry</span><span>)</span><span>type_clear</span><span>,</span>                        <span>/* tp_clear */</span>
    <span>0</span><span>,</span>                                          <span>/* tp_richcompare */</span>
    <span>offsetof</span><span>(</span><span>PyTypeObject</span><span>,</span> <span>tp_weaklist</span><span>),</span>        <span>/* tp_weaklistoffset */</span>
    <span>0</span><span>,</span>                                          <span>/* tp_iter */</span>
    <span>0</span><span>,</span>                                          <span>/* tp_iternext */</span>
    <span>type_methods</span><span>,</span>                               <span>/* tp_methods */</span>
    <span>type_members</span><span>,</span>                               <span>/* tp_members */</span>
    <span>type_getsets</span><span>,</span>                               <span>/* tp_getset */</span>
    <span>0</span><span>,</span>                                          <span>/* tp_base */</span>
    <span>0</span><span>,</span>                                          <span>/* tp_dict */</span>
    <span>0</span><span>,</span>                                          <span>/* tp_descr_get */</span>
    <span>0</span><span>,</span>                                          <span>/* tp_descr_set */</span>
    <span>offsetof</span><span>(</span><span>PyTypeObject</span><span>,</span> <span>tp_dict</span><span>),</span>            <span>/* tp_dictoffset */</span>
    <span>type_init</span><span>,</span>                                  <span>/* tp_init */</span>
    <span>0</span><span>,</span>                                          <span>/* tp_alloc */</span>
    <span>type_new</span><span>,</span>                                   <span>/* tp_new */</span>
    <span>PyObject_GC_Del</span><span>,</span>                            <span>/* tp_free */</span>
    <span>(</span><span>inquiry</span><span>)</span><span>type_is_gc</span><span>,</span>                        <span>/* tp_is_gc */</span>
<span>};</span>
</pre>
</div>
<p>The type of all built-in types is <code>type</code>, and the type of all classes defaults to <code>type</code>. So, <code>type</code> determines how types behave. For example, what happens when we call a type, like <code>list()</code> or <code>MyClass()</code>, is specified by the <code>tp_call</code> slot of <code>type</code>. The implementation of the <code>tp_call</code> slot of <code>type</code> is the <code>type_call()</code> function. Its job is to create new objects. It calls two other slots to do that:</p>
<ol>
<li>It calls <code>tp_new</code> of a type to create an object.</li>
<li>It calls <code>tp_init</code> of a type to initialize the created object.</li>
</ol>
<p>The type of <code>type</code> is <code>type</code> itself. So, when we call <code>type()</code>, the <code>type_call()</code> function is invoked. It checks for the special case when we pass a single argument to <code>type()</code>. In this case, <code>type_call()</code> simply returns the type of the passed object:</p>
<div>
<pre><span>$ python -q</span>
<span>&gt;&gt;&gt; </span><span>type</span><span>(</span><span>3</span><span>)</span>
<span>&lt;class 'int'&gt;</span>
<span>&gt;&gt;&gt; </span><span>type</span><span>(</span><span>int</span><span>)</span>
<span>&lt;class 'type'&gt;</span>
<span>&gt;&gt;&gt; </span><span>type</span><span>(</span><span>type</span><span>)</span>
<span>&lt;class 'type'&gt;</span>
</pre>
</div>
<p>But when we pass three arguments to <code>type()</code>, <code>type_call()</code> creates a new type by calling <code>tp_new</code> and <code>tp_init</code> of <code>type</code> as described above. The following example demonstrates how to use <code>type()</code> to create a class:</p>
<div>
<pre><span>$ python -q</span>
<span>&gt;&gt;&gt; </span><span>MyClass</span> <span>=</span> <span>type</span><span>(</span><span>'MyClass'</span><span>,</span> <span>(),</span> <span>{</span><span>'__str__'</span><span>:</span> <span>lambda</span> <span>self</span><span>:</span> <span>'Hey!'</span><span>})</span>
<span>&gt;&gt;&gt; </span><span>instance_of_my_class</span> <span>=</span> <span>MyClass</span><span>()</span>
<span>&gt;&gt;&gt; </span><span>str</span><span>(</span><span>instance_of_my_class</span><span>)</span>
<span>Hey!</span>
</pre>
</div>
<p>The arguments we pass to <code>type()</code> are:</p>
<ol>
<li>the name of a class</li>
<li>a tuple of its bases; and</li>
<li>a namespace.</li>
</ol>
<p>Other metatypes take arguments in this form as well.</p>
<p>We saw that we can create a class by calling <code>type()</code>, but that's not what we typically do. Typically, we use the <code>class</code> statement to define a class. It turns out that in this case too the VM eventually calls some metatype, and most often it calls <code>type()</code>.</p>
<p>To execute the <code>class</code> statement, the VM calls the <code>__build_class__()</code> function from the <code>builtins</code> module. What this function does can be summarized as follows:</p>
<ol>
<li>Decide which metatype to call to create the class.</li>
<li>
<a href="https://docs.python.org/3/reference/datamodel.html#preparing-the-class-namespace">Prepare</a> the namespace. The namespace will be used as a class's dictionary.</li>
<li>Execute the body of the class in the namespace, thus filling the namespace.</li>
<li>Call the metatype.</li>
</ol>
<p>We can instruct <code>__build_class__()</code> which metatype it should call using the <code>metaclass</code> keyword. If no <code>metaclass</code> is specified, <code>__build_class__()</code> calls <code>type()</code> by default. It also takes metatypes of bases into account. The exact logic of choosing the metatype is nicely <a href="https://docs.python.org/3/reference/datamodel.html#determining-the-appropriate-metaclass">described</a> in the docs.</p>
<p>Suppose we define a new class and do not specify <code>metaclass</code>. Where does the class actually get created? In this case, <code>__build_class__()</code> calls <code>type()</code>. This invokes the <code>type_call()</code> function that, in turn, calls the <code>tp_new</code> and <code>tp_init</code> slots of <code>type</code>. The <code>tp_new</code> slot of <code>type</code> points to the <code>type_new()</code> function. This is the function that creates classes. The <code>tp_init</code> slot of <code>type</code> points to the function that does nothing, so all the work is done by <code>type_new()</code>.</p>
<p>The <code>type_new()</code> function is nearly 500 lines long and probably deserves a separate post. Its essence, though, can be briefly summarized as follows:</p>
<ol>
<li>Allocate new type object.</li>
<li>Set up the allocated type object.</li>
</ol>
<p>To accomplish the first step, <code>type_new()</code> must allocate an instance of <code>PyTypeObject</code> as well as suites. Suites must be allocated separately from <code>PyTypeObject</code> because <code>PyTypeObject</code> contains only pointers to suites, not suites themselves. To handle this inconvenience, <code>type_new()</code> allocates an instance of the <code>PyHeapTypeObject</code> struct that extends <code>PyTypeObject</code> and contains the suites:</p>
<div>
<pre><span>/* The *real* layout of a type object when allocated on the heap */</span>
<span>typedef</span> <span>struct</span> <span>_heaptypeobject</span> <span>{</span>
    <span>PyTypeObject</span> <span>ht_type</span><span>;</span>
    <span>PyAsyncMethods</span> <span>as_async</span><span>;</span>
    <span>PyNumberMethods</span> <span>as_number</span><span>;</span>
    <span>PyMappingMethods</span> <span>as_mapping</span><span>;</span>
    <span>PySequenceMethods</span> <span>as_sequence</span><span>;</span>
    <span>PyBufferProcs</span> <span>as_buffer</span><span>;</span>
    <span>PyObject</span> <span>*</span><span>ht_name</span><span>,</span> <span>*</span><span>ht_slots</span><span>,</span> <span>*</span><span>ht_qualname</span><span>;</span>
    <span>struct</span> <span>_dictkeysobject</span> <span>*</span><span>ht_cached_keys</span><span>;</span>
    <span>PyObject</span> <span>*</span><span>ht_module</span><span>;</span>
    <span>/* here are optional user slots, followed by the members. */</span>
<span>}</span> <span>PyHeapTypeObject</span><span>;</span>
</pre>
</div>
<p>To set up a type object means to set up its slots. This is what <code>type_new()</code> does for the most part.</p>
<h2>Type initialization</h2>
<p>Before any type can be used, it should be initialized with the <code>PyType_Ready()</code> function. For a class, <code>PyType_Ready()</code> is called by <code>type_new()</code>. For a statically defined type, <code>PyType_Ready()</code> must be called explicitly. When CPython starts, it calls <code>PyType_Ready()</code> for each built-in type.</p>
<p>The <code>PyType_Ready()</code> function does a number of things. For example, it does slot inheritance.</p>
<h2>Slot inheritance</h2>
<p>When we define a class that inherits from other type, we expect the class to inherit some behavior of that type. For example, when we define a class that inherits from <code>int</code>, we expect it to support the addition:</p>
<div>
<pre><span>$ python -q</span>
<span>&gt;&gt;&gt; </span><span>class</span> <span>MyInt</span><span>(</span><span>int</span><span>):</span>
<span>... </span>    <span>pass</span>
<span>... </span>
<span>&gt;&gt;&gt; </span><span>x</span> <span>=</span> <span>MyInt</span><span>(</span><span>2</span><span>)</span>
<span>&gt;&gt;&gt; </span><span>y</span> <span>=</span> <span>MyInt</span><span>(</span><span>4</span><span>)</span>
<span>&gt;&gt;&gt; </span><span>x</span> <span>+</span> <span>y</span>
<span>6</span>
</pre>
</div>
<p>Does <code>MyInt</code> inherit the <code>nb_add</code> slot of <code>int</code>? Yes, it does. It's pretty straightforward to inherit the slots from a single ancestor: just copy those slots that the class doesn't have. It's a little bit more complicated when a class has multiple bases. Since bases, in turn, may inherit from other types, all these ancestor types combined form an hierarchy. The problem with the hierarchy is that it doesn't specify the order of inheritance. To solve this problem, <code>PyType_Ready()</code> converts this hierarchy into a list. <a href="https://www.python.org/download/releases/2.3/mro/">The Method Resolution Order</a> (MRO) determines how to perform this conversion. Once the MRO is calculated, it becomes easy to implement the inheritance in the general case. The <code>PyType_Ready()</code> function iterates over ancestors according to the MRO. From each ancestor, it copies those slots that haven't been set on the type before. Some slots support the inheritance and some don't. You can check in <a href="https://docs.python.org/3/c-api/typeobj.html#type-objects">the docs</a> whether a particular slot is inherited.</p>
<p>In contrast to a class, a statically defined type can specify at most one base. This is done by implementing the <code>tp_base</code> slot.</p>
<p>If no bases are specified, <code>PyType_Ready()</code> assumes that the <code>object</code> type is the only base. Every type directly or indirectly inherits from <code>object</code>. Why? Because it implements the slots that every type is expected to have. For example, it implements <code>tp_alloc</code>, <code>tp_init</code> and <code>tp_repr</code> slots.</p>
<h2>The ultimate question</h2>
<p>So far we've seen two ways in which a slot can be set:</p>
<ul>
<li>It can be specified explicitly (if a type is a statically defined type).</li>
<li>It can be inherited from an ancestor.</li>
</ul>
<p>It's still unclear how slots of a class are connected to its special methods. Moreover, we have a reverse problem for built-in types. How do they implement special methods? They certainly do:</p>
<div>
<pre><span>$ python -q</span>
<span>&gt;&gt;&gt; </span><span>(</span><span>3</span><span>)</span><span>.</span><span>__add__</span><span>(</span><span>4</span><span>)</span>
<span>7</span>
</pre>
</div>
<p>We come to the ultimate question of this post: What's the connection between special methods and slots?</p>
<h2>Special methods and slots</h2>
<p>The answer lies in the fact that CPython keeps a mapping between special methods and slots. This mapping is represented by the <code>slotdefs</code> array. It looks like this:</p>
<div>
<pre><span>#define TPSLOT(NAME, SLOT, FUNCTION, WRAPPER, DOC) \</span>
<span> {NAME, offsetof(PyTypeObject, SLOT), (void *)(FUNCTION), WRAPPER, \</span>
<span> PyDoc_STR(DOC)}</span>

<span>static</span> <span>slotdef</span> <span>slotdefs</span><span>[]</span> <span>=</span> <span>{</span>
    <span>TPSLOT</span><span>(</span><span>"__getattribute__"</span><span>,</span> <span>tp_getattr</span><span>,</span> <span>NULL</span><span>,</span> <span>NULL</span><span>,</span> <span>""</span><span>),</span>
    <span>TPSLOT</span><span>(</span><span>"__getattr__"</span><span>,</span> <span>tp_getattr</span><span>,</span> <span>NULL</span><span>,</span> <span>NULL</span><span>,</span> <span>""</span><span>),</span>
    <span>TPSLOT</span><span>(</span><span>"__setattr__"</span><span>,</span> <span>tp_setattr</span><span>,</span> <span>NULL</span><span>,</span> <span>NULL</span><span>,</span> <span>""</span><span>),</span>
    <span>TPSLOT</span><span>(</span><span>"__delattr__"</span><span>,</span> <span>tp_setattr</span><span>,</span> <span>NULL</span><span>,</span> <span>NULL</span><span>,</span> <span>""</span><span>),</span>
    <span>TPSLOT</span><span>(</span><span>"__repr__"</span><span>,</span> <span>tp_repr</span><span>,</span> <span>slot_tp_repr</span><span>,</span> <span>wrap_unaryfunc</span><span>,</span>
           <span>"__repr__($self, /)</span><span>\n</span><span>--</span><span>\n\n</span><span>Return repr(self)."</span><span>),</span>
    <span>TPSLOT</span><span>(</span><span>"__hash__"</span><span>,</span> <span>tp_hash</span><span>,</span> <span>slot_tp_hash</span><span>,</span> <span>wrap_hashfunc</span><span>,</span>
           <span>"__hash__($self, /)</span><span>\n</span><span>--</span><span>\n\n</span><span>Return hash(self)."</span><span>),</span>
    <span>// ... more slotdefs</span>
<span>}</span>
</pre>
</div>
<p>Each entry of this array is a <code>slotdef</code> struct:</p>
<div>
<pre><span>// typedef struct wrapperbase slotdef;</span>

<span>struct</span> <span>wrapperbase</span> <span>{</span>
    <span>const</span> <span>char</span> <span>*</span><span>name</span><span>;</span>
    <span>int</span> <span>offset</span><span>;</span>
    <span>void</span> <span>*</span><span>function</span><span>;</span>
    <span>wrapperfunc</span> <span>wrapper</span><span>;</span>
    <span>const</span> <span>char</span> <span>*</span><span>doc</span><span>;</span>
    <span>int</span> <span>flags</span><span>;</span>
    <span>PyObject</span> <span>*</span><span>name_strobj</span><span>;</span>
<span>};</span>
</pre>
</div>
<p>Four members of this struct are important for our discussion:</p>
<ul>
<li>
<code>name</code> is a name of a special method.</li>
<li>
<code>offset</code> is an offset of a slot in the <code>PyHeapTypeObject</code> struct. It specifies the slot corresponding to the special method.</li>
<li>
<code>function</code> is an implementation of a slot. When a special method is defined, the corresponding slot is set to <code>function</code>. Typically, <code>function</code> calls special methods to do the work.</li>
<li>
<code>wrapper</code> is a wrapper function around a slot. When a slot is defined, <code>wrapper</code> provides an implementation for the corresponding special method. It calls the slot to do the work.</li>
</ul>
<p>Here's, for example, an entry that maps <code>__add__()</code> special method to the <code>nb_add</code> slot:</p>
<ul>
<li>
<code>name</code> is <code>"__add__"</code>.</li>
<li>
<code>offset</code> is <code>offsetof(PyHeapTypeObject, as_number.nb_add)</code>.</li>
<li>
<code>function</code> is <code>slot_nb_add()</code>.</li>
<li>
<code>wrapper</code> is <code>wrap_binaryfunc_l()</code>.</li>
</ul>
<p>The <code>slotdefs</code> array is a many-to-many mapping. For example, as we'll see, both the <code>__add__()</code> and <code>__radd__()</code> special methods map to the same <code>nb_add</code> slot. Conversely, both the <code>mp_subscript</code> "mapping" slot and the <code>sq_item</code> "sequence" slot map to the same <code>__getitem__()</code> special method.</p>
<p>CPython uses the <code>slotdefs</code> array in two ways:</p>
<ul>
<li>to set slots based on special methods; and</li>
<li>to set special methods based on slots.</li>
</ul>
<h2>Slots based on special methods</h2>
<p>The <code>type_new()</code> function calls <code>fixup_slot_dispatchers()</code> to set slots based on special methods. The <code>fixup_slot_dispatchers()</code> function calls <code>update_one_slot()</code> for each slot in the <code>slotdefs</code> array, and <code>update_one_slot()</code> sets the slot to <code>function</code> if a class has the corresponding special method.</p>
<p>Let's take the <code>nb_add</code> slot as an example. The <code>slotdefs</code> array has two entries corresponding to that slot:</p>
<div>
<pre><span>static</span> <span>slotdef</span> <span>slotdefs</span><span>[]</span> <span>=</span> <span>{</span>
    <span>// ...</span>
    <span>BINSLOT</span><span>(</span><span>"__add__"</span><span>,</span> <span>nb_add</span><span>,</span> <span>slot_nb_add</span><span>,</span> <span>"+"</span><span>),</span>
    <span>RBINSLOT</span><span>(</span><span>"__radd__"</span><span>,</span> <span>nb_add</span><span>,</span> <span>slot_nb_add</span><span>,</span><span>"+"</span><span>),</span>
    <span>// ...</span>
<span>}</span>
</pre>
</div>
<p><code>BINSLOT()</code> and <code>RBINSLOT()</code> are macros. Let's expand them:</p>
<div>
<pre><span>static</span> <span>slotdef</span> <span>slotdefs</span><span>[]</span> <span>=</span> <span>{</span>
    <span>// ...</span>
    <span>// {name, offset, function,</span>
    <span>// wrapper, doc}</span>
    <span>// </span>
    <span>{</span><span>"__add__"</span><span>,</span> <span>offsetof</span><span>(</span><span>PyHeapTypeObject</span><span>,</span> <span>as_number</span><span>.</span><span>nb_add</span><span>),</span> <span>(</span><span>void</span> <span>*</span><span>)(</span><span>slot_nb_add</span><span>),</span>
        <span>wrap_binaryfunc_l</span><span>,</span> <span>PyDoc_STR</span><span>(</span><span>"__add__"</span> <span>"($self, value, /)</span><span>\n</span><span>--</span><span>\n\n</span><span>Return self"</span> <span>"+"</span> <span>"value."</span><span>)},</span>

    <span>{</span><span>"__radd__"</span><span>,</span> <span>offsetof</span><span>(</span><span>PyHeapTypeObject</span><span>,</span> <span>as_number</span><span>.</span><span>nb_add</span><span>),</span> <span>(</span><span>void</span> <span>*</span><span>)(</span><span>slot_nb_add</span><span>),</span>
        <span>wrap_binaryfunc_r</span><span>,</span> <span>PyDoc_STR</span><span>(</span><span>"__radd__"</span> <span>"($self, value, /)</span><span>\n</span><span>--</span><span>\n\n</span><span>Return value"</span> <span>"+"</span> <span>"self."</span><span>)},</span>
    <span>// ...</span>
<span>}</span>
</pre>
</div>
<p>What <code>update_one_slot()</code> does is look up <code>class.__add__()</code> and <code>class.__radd__()</code>. If either is defined, it sets <code>nb_add</code> of the class to <code>slot_nb_add()</code>. Note that both entries agree on <code>slot_nb_add()</code> as <code>function</code>. Otherwise, we would have a conflict when both are defined.</p>
<p>Now, what is <code>slot_nb_add()</code>, you ask? This function is defined with a macro that expands as follows:</p>
<div>
<pre><span>static</span> <span>PyObject</span> <span>*</span>
<span>slot_nb_add</span><span>(</span><span>PyObject</span> <span>*</span><span>self</span><span>,</span> <span>PyObject</span> <span>*</span><span>other</span><span>)</span> <span>{</span>
    <span>PyObject</span><span>*</span> <span>stack</span><span>[</span><span>2</span><span>];</span>
    <span>PyThreadState</span> <span>*</span><span>tstate</span> <span>=</span> <span>_PyThreadState_GET</span><span>();</span>
    <span>_Py_static_string</span><span>(</span><span>op_id</span><span>,</span> <span>"__add__"</span><span>);</span>
    <span>_Py_static_string</span><span>(</span><span>rop_id</span><span>,</span> <span>"__radd__"</span><span>);</span>
    <span>int</span> <span>do_other</span> <span>=</span> <span>!</span><span>Py_IS_TYPE</span><span>(</span><span>self</span><span>,</span> <span>Py_TYPE</span><span>(</span><span>other</span><span>))</span> <span>&amp;&amp;</span> \
        <span>Py_TYPE</span><span>(</span><span>other</span><span>)</span><span>-&gt;</span><span>tp_as_number</span> <span>!=</span> <span>NULL</span> <span>&amp;&amp;</span> \
        <span>Py_TYPE</span><span>(</span><span>other</span><span>)</span><span>-&gt;</span><span>tp_as_number</span><span>-&gt;</span><span>nb_add</span> <span>==</span> <span>slot_nb_add</span><span>;</span>
    <span>if</span> <span>(</span><span>Py_TYPE</span><span>(</span><span>self</span><span>)</span><span>-&gt;</span><span>tp_as_number</span> <span>!=</span> <span>NULL</span> <span>&amp;&amp;</span> \
        <span>Py_TYPE</span><span>(</span><span>self</span><span>)</span><span>-&gt;</span><span>tp_as_number</span><span>-&gt;</span><span>nb_add</span> <span>==</span> <span>slot_nb_add</span><span>)</span> <span>{</span>
        <span>PyObject</span> <span>*</span><span>r</span><span>;</span>
        <span>if</span> <span>(</span><span>do_other</span> <span>&amp;&amp;</span> <span>PyType_IsSubtype</span><span>(</span><span>Py_TYPE</span><span>(</span><span>other</span><span>),</span> <span>Py_TYPE</span><span>(</span><span>self</span><span>)))</span> <span>{</span>
            <span>int</span> <span>ok</span> <span>=</span> <span>method_is_overloaded</span><span>(</span><span>self</span><span>,</span> <span>other</span><span>,</span> <span>&amp;</span><span>rop_id</span><span>);</span>
            <span>if</span> <span>(</span><span>ok</span> <span>&lt;</span> <span>0</span><span>)</span> <span>{</span>
                <span>return</span> <span>NULL</span><span>;</span>
            <span>}</span>
            <span>if</span> <span>(</span><span>ok</span><span>)</span> <span>{</span>
                <span>stack</span><span>[</span><span>0</span><span>]</span> <span>=</span> <span>other</span><span>;</span>
                <span>stack</span><span>[</span><span>1</span><span>]</span> <span>=</span> <span>self</span><span>;</span>
                <span>r</span> <span>=</span> <span>vectorcall_maybe</span><span>(</span><span>tstate</span><span>,</span> <span>&amp;</span><span>rop_id</span><span>,</span> <span>stack</span><span>,</span> <span>2</span><span>);</span>
                <span>if</span> <span>(</span><span>r</span> <span>!=</span> <span>Py_NotImplemented</span><span>)</span>
                    <span>return</span> <span>r</span><span>;</span>
                <span>Py_DECREF</span><span>(</span><span>r</span><span>);</span> <span>do_other</span> <span>=</span> <span>0</span><span>;</span>
            <span>}</span>
        <span>}</span>
        <span>stack</span><span>[</span><span>0</span><span>]</span> <span>=</span> <span>self</span><span>;</span>
        <span>stack</span><span>[</span><span>1</span><span>]</span> <span>=</span> <span>other</span><span>;</span>
        <span>r</span> <span>=</span> <span>vectorcall_maybe</span><span>(</span><span>tstate</span><span>,</span> <span>&amp;</span><span>op_id</span><span>,</span> <span>stack</span><span>,</span> <span>2</span><span>);</span>
        <span>if</span> <span>(</span><span>r</span> <span>!=</span> <span>Py_NotImplemented</span> <span>||</span> <span>Py_IS_TYPE</span><span>(</span><span>other</span><span>,</span> <span>Py_TYPE</span><span>(</span><span>self</span><span>)))</span>
            <span>return</span> <span>r</span><span>;</span>
        <span>Py_DECREF</span><span>(</span><span>r</span><span>);</span>
    <span>}</span>
    <span>if</span> <span>(</span><span>do_other</span><span>)</span> <span>{</span>
        <span>stack</span><span>[</span><span>0</span><span>]</span> <span>=</span> <span>other</span><span>;</span>
        <span>stack</span><span>[</span><span>1</span><span>]</span> <span>=</span> <span>self</span><span>;</span>
        <span>return</span> <span>vectorcall_maybe</span><span>(</span><span>tstate</span><span>,</span> <span>&amp;</span><span>rop_id</span><span>,</span> <span>stack</span><span>,</span> <span>2</span><span>);</span>
    <span>}</span>
    <span>Py_RETURN_NOTIMPLEMENTED</span><span>;</span>
<span>}</span>
</pre>
</div>
<p>You don't need to study this code carefully. Recall the <code>binary_op1()</code> function that calls the <code>nb_add</code> slot. The <code>slot_nb_add()</code> function basically repeats the logic of <code>binary_op1()</code>. The main difference is that <code>slot_nb_add()</code> eventually calls <code>__add__()</code> or <code>__radd__()</code>.</p>
<h2>Setting special method on existing class</h2>
<p>Suppose that we create a class without the <code>__add__()</code> and <code>__radd__()</code> special methods. In this case, the <code>nb_add</code> slot of the class is set to <code>NULL</code>. As expected, we cannot add instances of that class. If we, however, set <code>__add__()</code> or <code>__radd__()</code> after the class has been created, the addition works as if the method was a part of the class definition. Here's what I mean:</p>
<div>
<pre><span>$ python -q</span>
<span>&gt;&gt;&gt; </span><span>class</span> <span>A</span><span>:</span>
<span>... </span>    <span>pass</span>
<span>... </span>
<span>&gt;&gt;&gt; </span><span>x</span> <span>=</span> <span>A</span><span>()</span>
<span>&gt;&gt;&gt; </span><span>x</span> <span>+</span> <span>2</span>
<span>Traceback (most recent call last):</span>
  File <span>"&lt;stdin&gt;"</span>, line <span>1</span>, in <span>&lt;module&gt;</span>
<span>TypeError</span>: <span>unsupported operand type(s) for +: 'A' and 'int'</span>
<span>&gt;&gt;&gt; </span><span>A</span><span>.</span><span>__add__</span> <span>=</span> <span>lambda</span> <span>self</span><span>,</span> <span>o</span><span>:</span> <span>5</span>
<span>&gt;&gt;&gt; </span><span>x</span> <span>+</span> <span>2</span>
<span>5</span>
<span>&gt;&gt;&gt; </span>
</pre>
</div>
<p>How does that work? To set an attribute on an object, the VM calls the <code>tp_setattro</code> slot of the object's type. The <code>tp_setattro</code> slot of <code>type</code> points to the <code>type_setattro()</code> function, so, when we set an attribute on a class, this function gets called. To set an attribute, it stores the value of the attribute in the class's dictionary. After this, it checks if the attribute is a special method. If it is a special method, the corresponding slots are set. The same <code>update_one_slot()</code> function is called to do that.</p>
<p>Before we can learn how CPython does the reverse, that is, how it adds special methods to built-in types, we need to understand what a method is.</p>
<h2>Methods</h2>
<p>A method is an attribute, but a peculiar one. When we call a method from an instance, the method implicitly receives the instance as its first parameter, which we usually denote <code>self</code>:</p>
<div>
<pre><span>$ python -q</span>
<span>&gt;&gt;&gt; </span><span>class</span> <span>A</span><span>:</span>
<span>... </span>    <span>def</span> <span>method</span><span>(</span><span>self</span><span>,</span> <span>x</span><span>):</span>
<span>... </span>            <span>return</span> <span>self</span><span>,</span> <span>x</span>
<span>...</span>
<span>&gt;&gt;&gt; </span><span>a</span> <span>=</span> <span>A</span><span>()</span>
<span>&gt;&gt;&gt; </span><span>a</span><span>.</span><span>method</span><span>(</span><span>1</span><span>)</span>
<span>(&lt;__main__.A object at 0x10d10bfd0&gt;, 1)</span>
</pre>
</div>
<p>But when we call the same method from a class, we have to pass all arguments explicitly:</p>
<div>
<pre><span>&gt;&gt;&gt; </span><span>A</span><span>.</span><span>method</span><span>(</span><span>a</span><span>,</span> <span>1</span><span>)</span>
<span>(&lt;__main__.A object at 0x10d10bfd0&gt;, 1)</span>
</pre>
</div>
<p>In our example, the method takes one argument in one case and two arguments in another. How is that possible that the same attribute is a different thing depending on how we access it?</p>
<p>First of all, realize that a method we define on a class is just a function. A function accessed through an instance differs from the same function accessed through the instance's type because the <code>function</code> type implements <a href="https://docs.python.org/3/howto/descriptor.html#descriptor-protocol">the descriptor protocol</a>. If you're unfamiliar with descriptors, I highly recommend you to read <a href="https://docs.python.org/3/howto/descriptor.html">Descriptor HowTo Guide</a> by Raymond Hettinger. In a nutshell, a descriptor is an object that, when used as an attribute, determines by itself how you get, set and delete it. Technically, a descriptor is an object that implements <a href="https://docs.python.org/3/reference/datamodel.html#object.__get__"><code>__get__()</code></a>, <a href="https://docs.python.org/3/reference/datamodel.html#object.__set__"><code>__set__()</code></a>, or <a href="https://docs.python.org/3/reference/datamodel.html#object.__delete__"><code>__delete__()</code></a> special methods.</p>
<p>The <code>function</code> type implements <code>__get__()</code>. When we look up some method, what we get is the result of a call to <code>__get__()</code>. Three arguments are passed to it:</p>
<ul>
<li>an attribute, i.e. a function</li>
<li>an instance</li>
<li>the instance's type.</li>
</ul>
<p>If we look up a method on a type, the instance is <code>NULL</code>, and <code>__get__()</code> simply returns the function. If we look up a method on an instance, <code>__get__()</code> returns a method object:</p>
<div>
<pre><span>&gt;&gt;&gt; </span><span>type</span><span>(</span><span>A</span><span>.</span><span>method</span><span>)</span>
<span>&lt;class 'function'&gt;</span>
<span>&gt;&gt;&gt; </span><span>type</span><span>(</span><span>a</span><span>.</span><span>method</span><span>)</span>
<span>&lt;class 'method'&gt;</span>
</pre>
</div>
<p>A method object stores a function and an instance. When called, it prepends the instance to the list of arguments and calls the function.</p>
<p>Now we're ready to tackle the last question.</p>
<h2>Special methods based on slots</h2>
<p>Recall the <code>PyType_Ready()</code> function that initializes types and does slot inheritance. It also adds special methods to a type based on the implemented slots. <code>PyType_Ready()</code> calls <code>add_operators()</code> to do that. The <code>add_operators()</code> function iterates over the entries in the <code>slotdefs</code> array. For each entry, it checks whether the special method specified by the entry should be added to the type's dictionary. A special method is added if it's not already defined and if the type implements the slot specified by the entry. For example, if the <code>__add__()</code> special method is not defined on a type, but the type implements the <code>nb_add</code> slot, <code>add_operators()</code> puts <code>__add__()</code> in the type's dictionary.</p>
<p>What is <code>__add__()</code> set to? Like any other method, it must be set to some descriptor to behave like a method. While methods defined by a programmer are functions, methods set by <code>add_operators()</code> are wrapper descriptors. A wrapper descriptor is a descriptor that stores two things:</p>
<ul>
<li>It stores a wrapped slot. A wrapped slot "does the work" for a special method. For example, the wrapper descriptor of the <code>__add__()</code> special method of the <code>float</code> type stores <code>float_add()</code> as a wrapped slot.</li>
<li>It stores a wrapper function. A wrapper function "knows" how to call the wrapped slot. It is <code>wrapper</code> of a <code>slotdef</code> entry.</li>
</ul>
<p>When we call a special method that was added by <code>add_operators()</code>, we call a wrapper descriptor. When we call a wrapper descriptor, it calls a wrapper function. A wrapper descriptor passes to a wrapper function the same arguments that we pass to a special methods plus the wrapped slot. Finally, the wrapper function calls the wrapped slot.</p>
<p>Let's see how a built-in type that implements the <code>nb_add</code> slot gets its <code>__add__()</code> and <code>__radd__()</code> special methods. Recall the <code>slotdef</code> entries corresponding to <code>nb_add</code>:</p>
<div>
<pre><span>static</span> <span>slotdef</span> <span>slotdefs</span><span>[]</span> <span>=</span> <span>{</span>
    <span>// ...</span>
    <span>// {name, offset, function,</span>
    <span>// wrapper, doc}</span>
    <span>// </span>
    <span>{</span><span>"__add__"</span><span>,</span> <span>offsetof</span><span>(</span><span>PyHeapTypeObject</span><span>,</span> <span>as_number</span><span>.</span><span>nb_add</span><span>),</span> <span>(</span><span>void</span> <span>*</span><span>)(</span><span>slot_nb_add</span><span>),</span>
        <span>wrap_binaryfunc_l</span><span>,</span> <span>PyDoc_STR</span><span>(</span><span>"__add__"</span> <span>"($self, value, /)</span><span>\n</span><span>--</span><span>\n\n</span><span>Return self"</span> <span>"+"</span> <span>"value."</span><span>)},</span>

    <span>{</span><span>"__radd__"</span><span>,</span> <span>offsetof</span><span>(</span><span>PyHeapTypeObject</span><span>,</span> <span>as_number</span><span>.</span><span>nb_add</span><span>),</span> <span>(</span><span>void</span> <span>*</span><span>)(</span><span>slot_nb_add</span><span>),</span>
        <span>wrap_binaryfunc_r</span><span>,</span> <span>PyDoc_STR</span><span>(</span><span>"__radd__"</span> <span>"($self, value, /)</span><span>\n</span><span>--</span><span>\n\n</span><span>Return value"</span> <span>"+"</span> <span>"self."</span><span>)},</span>
    <span>// ...</span>
<span>}</span>
</pre>
</div>
<p>If a type implements the <code>nb_add</code> slot, <code>add_operators()</code> sets <code>__add__()</code> of the type to a wrapper descriptor with <code>wrap_binaryfunc_l()</code> as a wrapper function and <code>nb_add</code> as a wrapped slot. It similarly sets <code>__radd__()</code> of the type with one exception: a wrapper function is <code>wrap_binaryfunc_r()</code>.</p>
<p>Both <code>wrap_binaryfunc_l()</code> and <code>wrap_binaryfunc_r()</code> take two operands plus a wrapped slot as their parameters. The only difference is how they call the slot:</p>
<ul>
<li>
<code>wrap_binaryfunc_l(x, y, slot_func)</code> calls <code>slot_func(x, y)</code>
</li>
<li>
<code>wrap_binaryfunc_r(x, y, slot_func)</code> calls <code>slot_func(y, x)</code>.</li>
</ul>
<p>The result of this call is what we get when we call the special method.</p>
<h2>Summary</h2>
<p>Today we've demystified perhaps the most magical aspect of Python. We've learned that the behavior of a Python object is determined by the slots of the object's type. The slots of a statically defined type can be specified explicitly, and any type can inherit some slots from its ancestors. The real insight was that the slots of a class are set up automatically by CPython based on the defined special methods. CPython does the reverse too. It adds special methods to the type's dictionary if the type implements the corresponding slots.</p>
<p>We've learned a lot. Nevertheless, the Python object system is such a vast subject that at least as much remains to be covered. For example, we haven't really discussed how attributes work. This is what we're going to do next time.</p>
<p><br></p>
<p><em>If you have any questions, comments or suggestions, feel free to contact me at victor@tenthousandmeters.com</em></p>
</div>
</div>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via blog | TenThousandMeters.com<br>
<br>
December 14, 2020 at 02:50PM
